src/main/java/com/google/sps/data/Cache.java

```diff
@@ -1,90 +1,77 @@
+      cell.addPoint(dataPoint);
+      row.put(index.col, cell);
+      dataGrid.put(index.row, row);
     }
 
-    dataGrid = new GriddedData(processingGrid, resolution, nwCorner);
-    return dataGrid;
+    return convertToGriddedData(dataGrid, aqDataPointsPerDegree, swCorner, neCorner);
   }
 
   private ArrayList<AQDataPoint> getNSWGovData() throws Exception {
     NSWGovAdaptor adaptor = new NSWGovAdaptor();
     return adaptor.getAQIData();
   }
 
-  private GridIndex getGridIndex(Coordinates originCoords, Coordinates targetCoords, int resolution) {
-    /*
-    * Layout of coordinates:
-    * originCoords -------------------- sameLngCoords
-    *       |                                  |
-    *       |                                  |
-    *       |                                  |
-    * sameLatCoords -------------------- targetCoords
-    * 
-    * Note: Due to the curvature of the earth, the actual shape bounded by the four
-    * points may not be exactly rectangular, however it should be close enough
-    * for small distances.
-    */
-    Coordinates sameLatCoords = new Coordinates(originCoords.lng,targetCoords.lat);
-    Coordinates sameLngCoords = new Coordinates(targetCoords.lng,originCoords.lat);
-
-    double latDistance = haversineDistance(originCoords, sameLngCoords);
-    double lngDistance = haversineDistance(originCoords, sameLatCoords);
-
-    int col = (int) Math.floor(latDistance / resolution);
-    int row = (int) Math.floor(lngDistance / resolution);
-
+  private GridIndex getGridIndex(Coordinates targetCoords, int aqDataPointsPerDegree) {
+    int row = (int) Math.floor(targetCoords.lat * aqDataPointsPerDegree);
+    int col = (int) Math.floor(targetCoords.lng * aqDataPointsPerDegree);
     return new GridIndex(col,row);
   }
 
-  private double haversineDistance(Coordinates point1, Coordinates point2) {
-    final int R = 6371000; // Radius of earth in metres
-    double rlat1 = Math.toRadians(point1.lat);
-    double rlat2 = Math.toRadians(point2.lat);
-    double difflat = rlat2 - rlat1;
-    double difflng = Math.toRadians(point2.lng - point1.lng);
-    double a = Math.pow(Math.sin(difflat / 2.0), 2) + Math.cos(rlat1) * Math.cos(rlat2) * Math.pow(Math.sin(difflng / 2.0),2);
-    double c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
-    double distance = R * c;
-    return distance;
+  // Note: this function does not support viewports that cross the 180 degree latitude or the poles
+  private GriddedData convertToGriddedData(HashMap<Integer,HashMap<Integer,GridCell>> data, int aqDataPointsPerDegree, Coordinates swCorner, Coordinates neCorner) {
+    GriddedData grid = new GriddedData(aqDataPointsPerDegree);
+    GridIndex swIndex = getGridIndex(swCorner, aqDataPointsPerDegree);
+    GridIndex neIndex = getGridIndex(neCorner, aqDataPointsPerDegree);
+
+    for (HashMap.Entry<Integer, HashMap<Integer, GridCell>> rowEntry : data.entrySet()) {
```

Comment:
This scans the entire globe, in order to populate a small view of it. Can we more efficiently define the space to scan?

(eg, pre-initialise the bounds of the result, then iterate that, fetching only the required points from the cache?)

Reply:
Yep, done